kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/ExecWebSocketListener.java

```diff
@@ -90,33 +107,39 @@ public void close() {
 
     public void close(int code, String reason) {
         try {
-         closeExecutor();
-         closeWebSocket(code, reason);
+          cleanupOnClose.set(true);
+          closeWebSocket(code, reason);
         } finally {
           if (listener != null) {
             listener.onClose(code, reason);
           }
         }
     }
 
-    private void closeExecutor() {
-      if (!executorClosed.compareAndSet(false, true)) {
+  /**
+   * Performs the cleanup tasks:
+   * 1. closes the InputStream pumper
+   * 2. closes all internally managed closeables (piped streams).
+   *
+   * The order of these tasks can't change or its likely that the pumper will through errors,
+   * if the stream it uses closes before the pumper it self.
+   */
+  private void cleanUp() {
```

Comment:
I can confirm that this is indeed true. I had to call ExecWatch.close inside an ExecListener which is less than ideal (because close calls the listener again). Otherwise I've found it deadlocks if you read the stdout on the PipeInputStream and no data has flowed and the websocket behind closes.  For example if the exec causes stderr only and no stdout, but something is already reading on the stdout PipeInputStream. 
I think the PipeOutputStream that is connected to PipeInputStream should always be closed when the onClose method is called. The problem is that ExecWatch.close() closes both the Pipe Input and Output Stream. So it's not safe to call in all conditions.